GLTexture.Texture constructor with image parameter of type LinearAlgebra.Adjoint

mfandl opened the pull request:
# Description

Fixes #1320
Fixes #1795

The array constructor for `GLMakie.GLAbstraction.Texture` had a concrete `Array{T, NDim}` type which caused calls to `Texture(...)` with transposed `Float32` arrays to result in an error (see linked issues for the actual message and stacktrace). This is because `LinearAlgebra.Adjoint{Float32, Matrix{Float32}} <: Array` does not hold.

This change fixes an issue when trying to produce `heatmap` or `image` from `Float32` arrays. Using different precision floats worked fine because these are converted to 32 bit floats such as in `tex = Texture(el32convert(mat), minfilter = interp)` and the adjoint is dropped.

I ran tests, however, I do not know whether I updated all the necessary places, or whether I did not cause a performance regression by this or something. Any pointers at improvements are welcome. The comment above the changed line mentions abstract arrays as types that are expected to work so I supposed this was a mistake?

MWE of the issue:

```julia
using GLMakie
heatmap(rand(Float32, 2, 3)')
```

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] Added an entry in NEWS.md (for new features and breaking changes)
- [ ] Added or changed relevant sections in the documentation
- [ ] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

SimonDanisch commented on GLMakie/src/GLAbstraction/GLTexture.jl:
```diff
@@ -134,7 +134,7 @@ So Array{Real, 2} == Texture2D with 1D Colorant dimension
 Colors from Colors.jl should mostly work as well
 """
-Texture(image::Array{T, NDim}; kw_args...) where {T <: GLArrayEltypes, NDim} =
+Texture(image::AbstractArray{T, NDim}; kw_args...) where {T <: GLArrayEltypes, NDim} =
```
That's not correct...
`pointer(...)` is not guaranteed to be correctly implemented for all AbstractArrays, and might even return a garbage pointer if used on the wrong array type. A lot of AbstractArrays like view + adjoint even have pointer defined, which just point to the parent array, which is of course not what you want from a view/mapping.

You'll need something like `Texture(image::AbstractArray{T, NDim}) = Texture(convert(Array{T, NDIM}, image))` to be on the safe side.
I'm guessing this "fixed" your issue but the data must have been incorrectly permuted:
<img width="442" alt="image" src="https://user-images.githubusercontent.com/1010467/214568085-1eaa9a81-49a9-41ec-a9e1-ba5d7067e58a.png">

SimonDanisch commented:
Thank you! As pointed out in the review, this was not a mistake, the "abstractarray" refers to static arrays like `Vec2f` for the element type.
Can you give an MWE, which currently fails sicne it isn't running into `el32convert`?

mfandl commented:
@SimonDanisch Ah, thanks for the comments, I see why this is problematic now. Sorry for assuming that this was a mistake, I was just thinking that based on the comment in code.

> Can you give an MWE, which currently fails sicne it isn't running into `el32convert`?

I think the problem is not that `el32convert` wouldn't be called. It just acts as an identity function in the following case and therefore retains the currently unsupported type:

```julia
julia> x = rand(Float32, 2, 3)'
3×2 adjoint(::Matrix{Float32}) with eltype Float32:
 0.926551  0.0612232
 0.560767  0.139857
 0.617872  0.161374

julia> typeof(x)
LinearAlgebra.Adjoint{Float32, Matrix{Float32}}

julia> typeof(GLMakie.el32convert(x))
LinearAlgebra.Adjoint{Float32, Matrix{Float32}}
```

this is (understandably) different in case the element type is different:

```julia
julia> typeof(GLMakie.el32convert(rand(Float64, 2, 3)))
Matrix{Float32} (alias for Array{Float32, 2})
```

I thought this is actually correct behavior, that is why I considered updating the defintion to contain the `AbstractArray` type a proper solution.

Could you please help me understand what would be the the right direction? A solution along the lines of what you write above, or updating the behavior of el32 convert, or something else?

Anyway, the MWE of the problem I am trying to fix is this:

```julia
using GLMakie
heatmap(rand(Float32, 2, 3)')
```

I will also add it to the description of the PR.

mfandl commented:
I updated the commit taking the code review comments into account - I reverted the `AbstractArray` change and introduced a new constructor which accepts the type which caused problems and collects it.

asinghvi17 commented on GLMakie/src/GLAbstraction/GLTexture.jl:
```diff
@@ -137,6 +137,9 @@ Colors from Colors.jl should mostly work as well
 Texture(image::Array{T, NDim}; kw_args...) where {T <: GLArrayEltypes, NDim} =
     Texture(pointer(image), size(image); kw_args...)::Texture{T, NDim}
 
+Texture(image::Adjoint{T, Array{T, NDim}}; kw_args...) where {T <: GLArrayEltypes, NDim} =
```
should this be `image::AbstractArray{T, NDim}` for the most generality?

mfandl replied:
Thank you for your review.

Originally this PR updated the `image::Array{T, NDim}` constructor to exactly what you write. But @SimonDanisch shed light on why this is wrong in his code review above. Would keeping the old `image::Array{T, NDim}` and adding an extra `image::AbstractArray{T, NDim}` constructor where I'd `collect` whatever the concrete `AbstractArray` subtype is make things safe? (Actually, probably not `collect` in the more general case, but the proposal from the first review: `Texture(image::AbstractArray{T, NDim}) = Texture(convert(Array{T, NDIM}, image))` - but the question remains, is this safe to do for any `AbstractArray`?)

You are of course right that more could be fixed in this PR now that I think about it more, this is another example that throws error of the same kind:

```julia
using GLMakie
x = rand(Float32, 5, 5);
y = @view x[2:3, 2:3];
heatmap(y)
```
```
Error showing value of type Makie.FigureAxisPlot:
ERROR: MethodError: no method matching GLMakie.GLAbstraction.Texture(::SubArray{Float32, 2, Matrix{Float32}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}; minfilter=:nearest)
```
And the result of joining these two types really is an `AbstractArray`:

```
julia> typejoin(SubArray, Adjoint)
AbstractArray
```
